File: Leetcode/1372 Longest ZigZag Path in a Binary Tree.py
class Solution:
    def longestZigZag(self, root):
        self.max_len = 0
        self.dfs(root)
        return self.max_len
    
    def dfs(self, root):
        if root is None:
            return -1, -1
        _, right = self.dfs(root.left)
        left, _ = self.dfs(root.right)
        from_root = 1 + right, 1 + left
        self.max_len = max(self.max_len, *from_root)
        return from_root 

File: Leetcode/1372 Longest ZigZag Path in a Binary Tree.ts
function longestZigZag(root: TreeNode | null): number {
    let maxLen = 0;
    function dfs(root: TreeNode | null) {
        if (root === null) {
            return {left: -1, right: -1};
        }
        const left = dfs(root.left);
        const right = dfs(root.right);
        const fromRoot = {
            left: 1 + left.right,
            right: 1 + right.left,
        }
        maxLen = Math.max(maxLen, fromRoot.left, fromRoot.right);
        return fromRoot;
    }
    dfs(root);
    return maxLen;
}
